### PR TITLE
[Exam] Bugfix tutor dashboard

### PR DESCRIPTION
**Pull request by HanyaElhashemy**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Fixed following bugs:
1. Allow tutors to access the exam management and exam exercise dashboard from the course management view if the exam due date is over
3. (Tutor) There is a permission problem with example submissions (to review / assess): Tutors cannot access the example submissions
4. (Instructor) When creating modeling exercises, there is something about text exercises ("[...] must finish creating the text exercise" or so)
5. (Instructor) The back buttons for creating the Example Solution and the Example Submissions during modeling do not work (-> console error)
6. (Instructor, as Tutor does not have access yet) The back buttons for leaving the example submissions (to review / assess) in the Tutor Exam Dashboard do not work (-> console error)

### Steps for Testing

1. Navigate to course management
2. click on exams 
3. add an exercise group with a modeling exercise
4. check that the bugs are fixed -> play around with creating example solution/submission
5. check that you don't get console error when navigating back from the example solution/submission page
6. navigate to the tutor dashboard and check if you have permission to access example submissions
-----------------------------
1. login as a tutor only
2. navigate to exam management -> check that you *only* have access to the tutor dashboard
3. access the dashboard and try access an example submission

### CODE REVIEW

**Comment by pzdr7** [comment]
> navigate to the tutor dashboard and check if you have permission to access example submissions

We have tested this and we found that a regular tutor still does not have access to the example submissions.
This seems to be the case for modeling and text exercises alike:

Modeling:

image

Text:

image

**Comment by krusche** [comment]
We need two additional improvement:
1) Tutors should be able to navigate in the exam scores page (at least if they know the URL)
2) Here, the manual result button and the filters (manual vs. automatic) should be shown for exam exercises as well in case manual assessment is activated in the programming exercise

@filip-gregurevic @sascha11110 can you help to review and test the changes here and implement the additional improvement (1)

@anditurdiu can you implement the additional improvement (2)

File: src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.ts [hunk 1]
```diff
@@ -370,11 +370,15 @@ export class ExampleTextSubmissionComponent implements OnInit, AfterViewInit {
...
     async back() {
         // check if exam exercise
-        if (this.exercise.course == null) {
+        if (this.exercise.course == null || !this.exercise.course) {
             const courseId = this.exercise.exerciseGroup?.exam?.course.id;
             const examId = this.exercise.exerciseGroup?.exam?.id;
             const exerciseGroupId = this.exercise.exerciseGroup?.id;
-            await this.router.navigate(['/course-management', courseId, 'exams', examId, 'exercise-groups', exerciseGroupId, 'text-exercises', this.exerciseId, 'edit']);
+            if (this.readOnly || this.toComplete) {
+                this.router.navigate([`/course-management/${courseId}/exercises/${this.exerciseId}/tutor-dashboard`]);
```
Comment (sascha11110) [inline src/main/webapp/app/exercises/text/manage/example-text-submission/example-text-submission.component.ts:378]: To be consistent we should add `await` here

File: src/test/java/de/tum/in/www1/artemis/ExerciseIntegrationTest.java [hunk 2]
```diff
@@ -206,12 +206,6 @@ public void testGetExamExercise_asStudent_forbidden() throws Exception {
         getExamExercise();
     }
 
-    @Test
-    @WithMockUser(value = "tutor1", roles = "TA")
-    public void testGetExamExercise_asTutor_forbidden() throws Exception {
-        getExamExercise();
-    }
-
```
Comment (sascha11110) [inline src/test/java/de/tum/in/www1/artemis/ExerciseIntegrationTest.java:214]: Add explicit test case

File: src/test/java/de/tum/in/www1/artemis/service/ExamAccessServiceTest.java [hunk 3]
```diff
@@ -114,29 +114,6 @@ public void testCheckCourseAccess_asStudent() {
...
-                exerciseGroup1.getId());
-        assertThat(accessFailureCourseAndExamAndExerciseGroup.isPresent()).isTrue();
-        assertThat(accessFailureCourseAndExamAndExerciseGroup.get()).isEqualTo(forbidden());
-        // checkCourseAndExamAndStudentExamAccess
-        Optional<ResponseEntity<Exam>> accessFailureCourseAndExamAndStudentExam = examAccessService.checkCourseAndExamAndStudentExamAccess(course1.getId(), exam1.getId(),
-                studentExam1.getId());
-        assertThat(accessFailureCourseAndExamAndStudentExam.isPresent()).isTrue();
-        assertThat(accessFailureCourseAndExamAndStudentExam.get()).isEqualTo(forbidden());
-    }
-
```
Comment (sascha11110) [inline src/test/java/de/tum/in/www1/artemis/service/ExamAccessServiceTest.java:139]: Possibly not all of the routes should be open to tutors

**Comment by anditurdiu** [comment]
> We need two additional improvement:
> 
> 1. Tutors should be able to navigate in the exam scores page (at least if they know the URL)
> 2. Here, the manual result button and the filters (manual vs. automatic) should be shown for exam exercises as well in case manual assessment is activated in the programming exercise
> 
> @filip-gregurevic @sascha11110 can you help to review and test the changes here and implement the additional improvement (1)
> 
> @anditurdiu can you implement the additional improvement (2)

I fixed point 2, to test:
1. Create an exam with a programing exercise. Make sure to set the start time and end time as well as check the manual assessment (at the timeline, after due date).
2. Start the exam, make a submission 
3. Go to Course Management => Exams => Exercise Groups, click on the scores page of the programming exercise. 
4. You should now see the "New Result" button

image

**Comment by TobiasPr** [comment]
Still doing review - one thing I found: after creating sample solution + sample submission the save and cancel button do not work as expected (wrong navigation)
sample_solution_modeling_exercise
I think this also happens when creating an exercise for a course

**Comment by krusche** [comment]
> Still doing review - one thing I found: after creating sample solution + sample submission the save and cancel button do not work as expected (wrong navigation)
> sample_solution_modeling_exercise
> I think this also happens when creating an exercise for a course - I don't know if you want to tackle that in this PR or if we should open an issue for that

Yes, this is a general issue that I noticed as well. Can be fixed in a follow up

**Comment by JonasPetry** [comment]
I've navigated into a modeling assessment for the first time and clicking on "Start participation", I've got stuck on this page:

image

This error only occurs on the first "participation", when leaving this page and going back everything works again.

File: src/main/webapp/app/course/dashboards/tutor-course-dashboard/tutor-course-dashboard.component.html [hunk 4]
```diff
@@ -3,8 +3,8 @@
         <div class="col-8">
             <h2>
                 <fa-icon [icon]="'arrow-left'" (click)="back()" class="back-button mr-2"></fa-icon>
-                {{ 'artemisApp.tutorCourseDashboard.pageHeader' | translate }} {{ course?.title }}
-
+                <h2 *ngIf="!isExamMode">{{ 'artemisApp.tutorCourseDashboard.pageHeader' | translate }} {{ course?.title }}</h2>
+                <h2 *ngIf="isExamMode">{{ 'artemisApp.tutorCourseDashboard.pageHeaderExam' | translate }} {{ exam.title }}</h2>
```
Comment (JonasPetry) [inline src/main/webapp/app/course/dashboards/tutor-course-dashboard/tutor-course-dashboard.component.html:7]:
```suggestion
                <h2 *ngIf="isExamMode">{{ 'artemisApp.tutorCourseDashboard.pageHeaderExam' | translate }} {{ exam?.title }}</h2>
```

**Comment by JonasPetry** [comment]
> Can anyone confirm this bug? I've tested this on TS3.
> 
> 1. Sign in as an instructor
> 2. Create an exam
> 3. Create an exercise group, add a modeling exercise
> 4. Edit the modeling exercise, try to add an example submission ("Create Example Submission")
> 5. The page will remain empty, there will be a 403 in the console:
> 
> image
> 
> A similar problem presents itself (even with admin permissions) with text exercises. This error message occurs when hitting ("Create Example Submission")
> 
> image

I cannot confirm this bug. For me it is working with modeling exercises. With text exercises, I also have the "null" in the url, but everything works as expected.

**Comment by TobiasPr** [comment]
Something else happened to me when testing
1. The tutor for my course can create and edit exams -> is that correct?
2. When the exam was over and I navigated to the Tutor Exercise dashboard, I saw my sample submission, when I click on start new assessment, a warning was shown see screenshot
image
3. When I navigate into an assessment - navigating back / canceling requires two clicks (only when I start an assessment, not when I continue an assessment, can be reproduced by canceling assessment)
4. when I tried to navigate into text assessment I got an ChunkLoadError - I assume due to permissions
image

So maybe all my testing does not help anything because I assume my role somehow got mixed up even though I was using artemis_test_user_6 in the "Adrian Test Course" on ts3 in which test user 6 is not a instructor (I tested in chrome Incognito window)
image
image
image

**Comment by filip-gregurevic** [comment]
> Something else happened to me when testing
> 
> 1. The tutor for my course can create and edit exams -> is that correct?

As far as I can see in the screenshots your account has the 'admin' role, since the server management is visible in the header, that might be the reason

**Comment by TobiasPr** [comment]
> > Something else happened to me when testing
> > 
> > 1. The tutor for my course can create and edit exams -> is that correct?
> 
> As far as I can see in the screenshots your account has the 'admin' role, since the server management is visible in the header, that might be the reason

Okay then I will delete all things related to permission, since the user has the wrong roles on ts3

**Comment by julian-christl** [comment]
@TobiasPr Do you have added an assessment to your example submission?

**Comment by Varnika-Tyagi** [comment]
Tested it on TS3. It works as intended.
**Bugs fixed:** 

- _Allow tutors to access the exam management and exam exercise dashboard from the course management view if the exam due date is over_
Works. Tutors can only see the exam dashboard once the exam is over.
- _(Tutor) There is a permission problem with example submissions (to review / assess): Tutors cannot access the example submissions_
Works as intended. I was able to view example assessment and review example submissions, for both modeling and text exercises.
- _(Instructor) When creating modeling exercises, there is something about text exercises ("[...] must finish creating the text exercise" or so)_
Works fine. I didn't encounter any error while creating the modelling or text exercises, apart from the null in the url for text exercises, as @JonasPetry mentioned.
- _(Instructor) The back buttons for creating the Example Solution and the Example Submissions during modeling do not work (-> console error)_
They are working fine now. But the behavior of Save and Cancel buttons on the 'Create exercise' page seems to be working incorrectly. @TobiasPr and @krusche already mentioned it as out of scope of this pr
- _(Instructor, as Tutor does not have access yet) The back buttons for leaving the example submissions (to review / assess) in the Tutor Exam Dashboard do not work (-> console error)_
The buttons are working as intended.

Reviewing example submissions:
Example Submissions tutor dashboard
Text review assessment
Tutor assessment:
Assessing submissions

Well done!

Also, (out of scope here, but) the Exam Summary looks great now, with the modelling exercise submissions displayed correctly in the pdf.
Summary

**Pull request merged**